[server] server connections state tracking

mmatczuk opened the pull request:
In this PR we have introduced a server channel that allows for connection state tracking. The design is C&P form what is already done for client.

mmatczuk commented:
@rjeczalik gorilla does not longer compile with go 1.4 :(

mmatczuk commented:
BTW You are missing checks for 1.6

rjeczalik commented on server.go:
```diff
@@ -495,17 +505,56 @@ func (s *Server) listenControl(ct *control) {
+		s.log.Error("OnConnect: error calling callback for %q: %s", identifier, err)
+	}
+
+	if (s.stateChanges != nil) {
```
Could you run `gofmt -w -s .` over the files in this PR?

rjeczalik commented on server.go:
```diff
@@ -26,6 +26,11 @@ var (
 
+type ConnectionStateChange struct {
+	Identifier string
+	Connected  bool
```
`StateChange` is about state transition, this value carries information how the connection state transitioned. While I don't see for now use of tracking connection's previous state, can we add `Current ServerState` field here and get rid of `Connected` one? Then we could have `ServerConnected`, `ServerDisconnected` and in future it would be easier to add more fine-grained events like `ServerSessionRequested`, `ServerHeartbeat`, `ServerSessionRetry` etc. Something like:

```go
type ServerState uint8

type ServerStateChange struct {
   Identifier string
   Current  ServerState
}
```

I was also thinking how hard would it be to add events for abnormal disconnections - e.g. yamux transport timed out sending heartbeat or requesting session and decided to drop connection; for such cases emitting `ServerStateChange` with `Error error` field set would be handful for e.g. tracking statistics. Just thinking loudly.

mmatczuk replied:
It used to be enum but I finally decided that 2 value enum is a little bit overdone.

:+1: I'll check what can I do for error propagation

rjeczalik commented:
> BTW You are missing checks for 1.6

Yeah, there's a bit more things missing - the ones I care the most about are benchmarks. Could you throw 1.6.2 here and bump 1.5.3 -> 1.5.4?
https://github.com/koding/tunnel/blob/master/.travis.yml#L5-L7

mmatczuk commented:
I pushed config update I'll get back to you later with error and enum.

mmatczuk-stratoscale commented:
@rjeczalik I just wanted to let you know that I remember about the pending pull requests. I'll work more on that in July.

mmatczuk commented:
@rjeczalik I just wanted to let you know that I remember about the pending pull requests. I'll work more on that in July.

rjeczalik commented:
@mmatczuk Thanks for heads up 👍 

Most likely I'm going to work on https://github.com/koding/tunnel/pull/23 - I need to extend the API to allow passing custom proxy functions, so I wanted to also cover your case with `FetchAddr`.

mmatczuk commented:
@rjeczalik I have implemented the proposed changes, namely:

1. the channel now uses a transition struct pointer
1. onDisconnect propagates errors

I also added onDisconnect call upon unsuccessful connect (before starting listenControl).

PR contains tests.

Regards

rjeczalik commented on client.go:
```diff
@@ -391,6 +392,7 @@ func (c *Client) changeState(state ClientState, err error) (prev ClientState) {
 		select {
 		case c.config.StateChanges <- change:
 		default:
+			c.log.Warning("Status change dropped")
```
Please add more context here:

```go
c.log.Warning("dropping state change due to slow reader: %s -> %s", change.Previous, change.Current)
```

rjeczalik commented on server.go:
```diff
@@ -69,6 +96,14 @@ type Server struct {
 
 // ServerConfig defines the configuration for the Server
 type ServerConfig struct {
+	// ClientStateChanges receives state transition details each time client
```
s/each time client/each time server/

rjeczalik commented on server.go:
```diff
@@ -69,6 +96,14 @@ type Server struct {
+	//
+	// If nil, no information about state transitions are dispatched
+	// by the library.
+	ClientStateChanges chan<- *ServerClientStateChange
```
ServerStateChange or just StateChanges

rjeczalik commented on server.go:
```diff
@@ -496,17 +531,68 @@ func (s *Server) listenControl(ct *control) {
+		select {
+		case s.clientStateCh <- change:
+		default:
+			s.log.Warning("Status change dropped")
```
The same as above.

rjeczalik commented:
@mmatczuk Looks good :+1:

mmatczuk commented:
@rjeczalik it was also my impression that naming is a bit awkward but I did not have a better idea yesterday.

How about we introduced Identifier to ClientStateChange and used that in Server as well. This would streamline the design and naming. In that case we would also use StateRecorder for server testing as well. It means that server would use only a subset of client states. What do you think?

Also I need to fix state transitions flow for consistency with client code and make sure unknown used only when client never connected before.

rjeczalik commented:
> How about we introduced Identifier to ClientStateChange and used that in Server as well.

I like this idea 👍

mmatczuk commented:
@rjeczalik I wonder why there are two testing packages tunnel_test and tunneltest. Could they be merged to tunnel_test. I want to replace TunnelTest.ClientStateChanges with tunnel_test.StateRecorder which introduces package dependency, so either

* packages are merged
* StateRecorder is moved to tunneltest
* or we add dependency

Do you have any opinions on that front?

rjeczalik commented:
>  I wonder why there are two testing packages tunnel_test and tunneltest.

To have testing code that may be reused in external packages. The *_test packages can't be imported from outside. Moving StateRecorder to tunneltest sounds good to me :-)

mmatczuk commented:
Ok

mmatczuk commented:
@rjeczalik I squashed all commits and rebased it to master.

1. ClientStateChange is used for Client and server,
1. StateRecorder is moved to tunneltest, 
1. tests are updated to use state recorder,
1. state transitions are fixed (no delete).

rjeczalik commented on server.go:
```diff
@@ -69,6 +79,14 @@ type Server struct {
+	//
+	// If nil, no information about state transitions are dispatched
+	// by the library.
+	StateChanges chan<- *ClientStateChange
```
Can I ask you one more thing? If we're going to use the same ClientStateChange for both server and client, how about renaming it to just `StateChange`?

mmatczuk-stratoscale replied:
@rjeczalik I see your point I left is for code backward compatibility.

mmatczuk commented:
@rjeczalik do you still want to do the rename?

rjeczalik commented:
@mmatczuk ops, sorry for making you wait - it's ok, backward-compatibility is a good argument for keeping the Client prefixes. Having to adapt the code for no real gain after updating deps can be tiresome.

Thanks!

mmatczuk-stratoscale commented:
@rjeczalik Thanks a lot for merging this change.

mmatczuk commented:
@rjeczalik Thanks a lot for merging this change.
